### PR TITLE
Save command-result card to workspace

### PR DESCRIPTION
**Pull request by burieberry**
Opened the menu with the button again after copy for the screenshot:
copy-to-workspace

- Added `copyCard` action in interact submode (because the logic for checking the realm of any open card in stack required this to live there)
- Made actions in interact-submode available in card context (using the `@provide` pattern)
- Added an option (`isWritableRealm`) to `getRealmURL` function in `cardService` for it to return a writable realm url. By default it returns a readable realm url.
- How it works: CommandCard and CommandResult card are saved in the matrix server, however we do not save them locally. If a user clicks on `Copy to Workspace`, then we save a copy of the CommandResult card in user's workspace.

Note:
- Each time the "Copy to Workspace" button is clicked, there'll be a new copy of the card saved. I think this is consistent with what the button is meant to do, but wanted to hear if people think otherwise.

### CODE REVIEW

**Review by tintinthong — COMMENTED** [review]
Is this supposed to show anything? the search result in edit mode

Screenshot 2024-10-07 at 21 17 43

**Comment by tintinthong** [comment]
I sometimes see results not appearing. I thought this was actually fixed before.. is it your experience to see this often?

**Comment by tintinthong** [comment]
So we copy the result card to the workspace but not the cards itself. The result card will just linksToMany back to wherever the cards come from?

**Comment by burieberry** [comment]
> I sometimes see results not appearing. I thought this was actually fixed before.. is it your experience to see this often?

Are you talking about CS-7255? I only saw that happen on prod. I haven't checked recently on prod.

**Comment by burieberry** [comment]
> So we copy the result card to the workspace but not the cards itself. The result card will just linksToMany back to wherever the cards come from?

Yes, the idea is that any card shared in chat can be copied to workspace the same way. In the isolated search results card, the results are clickable.

**Comment by burieberry** [comment]
> Is this supposed to show anything? the search result in edit mode
> 
> Screenshot 2024-10-07 at 21 17 43

Hmm, I'll take a look what we can do there.

File: packages/base/command.gts [hunk 1]
```diff
@@ -1,34 +1,16 @@
...
+class CommandStatusField extends StringField {
   static [primitive]: CommandStatus;
 }
 
-export class CommandField extends CardDef {
+export class CommandCard extends CardDef {
   @field toolCallId = contains(StringField);
   @field name = contains(StringField);
-  @field payload = contains(CommandObjectField); //arguments of toolCall. Its not called arguments due to lint
+  @field payload = contains(ToolCallArg); //arguments of toolCall. Its not called arguments due to lint
```
Comment (lukemelia) [inline packages/base/command.gts]: I think we prefer this to be more abstract than "ToolCallArg". That phrase is AI Bot specific as far as I understand it, and not all commands should interact with AI.

File: packages/host/app/components/operator-mode/interact-submode.gts [hunk 2]
```diff
@@ -623,6 +642,14 @@ export default class InteractSubmode extends Component<Signature> {
     openSearchCallback();
   }
 
+  @provide(CardContextName)
+  // @ts-ignore "cardContext is declared but not used"
+  private get cardContext() {
+    return {
+      actions: this.publicAPI(this, this.stacks[0]?.length ? 1 : 0),
```
Comment (lukemelia) [inline packages/host/app/components/operator-mode/interact-submode.gts]: I know that targeting the right column is the correct default for the copy command, but is it really the correct default for all actions?

File: packages/base/command.gts (hunk 1 above)
Comment (burieberry) [reply packages/base/command.gts]: is `CommandObjectField` a better name then? I also considered converting it to a string and parsing as needed.

File: packages/host/app/services/card-service.ts [hunk 3]
```diff
@@ -452,8 +452,18 @@ export default class CardService extends Service {
...
+        return url;
+      }
+      if (!this.realm.defaultWritableRealm) {
+        throw new Error('Could not find a writable realm');
+      }
+      return new URL(this.realm.defaultWritableRealm.path);
+    }
     // in the case where we get no realm URL from the card, we are dealing with
     // a new card instance that does not have a realm URL yet.
     return card[api.realmURL] ?? new URL(this.realm.defaultReadableRealm.path);
```
Comment (ef4) [inline packages/host/app/services/card-service.ts]: I see that this PR is extending the existing fallback behavior. But this fallback behavior doesn't belong here.

`getRealmURL` should only return a card's actual realm url. When there isn't one, it should return undefined, and it should be up to the caller to pick a default instead.

**Pull request merged**